### ISSUE TITLE
Optimize Paul's riddle KCL

### ISSUE THREAD

**Issue by adamchalmers** [KittyCAD/modeling-app#1969]
Paul's [riddle.kcl](https://github.com/KittyCAD/modeling-app/blob/main/src/wasm-lib/tests/executor/inputs/riddle.kcl) still takes WAY too much time to practically run, even with batching. That is because each call to `p` generates several batches.

I modified the program to only call `p` three times, and to print out every API request batch that got executed. Here's the [results](https://github.com/KittyCAD/modeling-app/files/14807680/riddle_batches.txt).

Some things jump out:

- Firstly, there's a LOT of `SetSceneUnits` calls being sent. The executor sets the scene to millimetres _ten times_ split into _six batches_ in the shortened program. This is bad for two reasons: it's unnecessary WebSocket roundtrips, and it might be forcing batch flushes.
- Each call to `p` creates its own set of batches -- ideally all the `p` calls would be combined into one batch, but that might not be possible.
- In each call to `p` the main batch starts with a "preamble" of `MakePlane, SketchModeEnable, MakePlane, SketchModeEnable` before it does the actual sketching work (i.e. `StartPath, MovePathPen, ExtendPath x3, ClosePath`). 
    - SketchModeEnable is not necessary here
    - I don't think MakePlane is necessary either. It's certainly not necessary _multiple times_. The MakePlane calls all have the same parameters, so they're just repeating unnecessary work).
    - At the end of this main batch, there's a call to `SketchModeDisable` which can also be skipped.
- After each "main batch" there's an Extrude, an ObjectBringToFront and a Solid3DGetExtrusionFaceInfo, all in their own individual batches. The Extrude and ObjectBringToFront can probably go in the main batch. 
- Solid3DGetExtrusionFaceInfo calls actually return data, so they can't really be batched easily. This means that every extrude has to happen in its own batch (as we need to query the extrude for the GUI). This means models with lots of extrudes are going to take a _very_ long time (1000 extrudes, each requiring 1 engine frame, at 60fps takes 16 seconds). One way to fix this is: stop querying the extrusion face info every time you extrude. Instead, after extruding, note the ID of the extruded face. Then at the end of the program, query all the extruded face info at once.

To summarize, I've filed these issues:

- [ ] https://github.com/KittyCAD/modeling-app/issues/1972
- [x] https://github.com/KittyCAD/modeling-app/issues/1974
- [ ] https://github.com/KittyCAD/modeling-app/issues/1973
- [x] https://github.com/KittyCAD/modeling-app/issues/1971

These could all be fixed with their own PRs.

**Comment by jessfraz** [comment]
closing as we have benchmarks and other things to optimize more now
